rust/noosphere/src/ffi/petname.rs

```diff
@@ -0,0 +1,47 @@
+#![allow(unused_variables)]
+
+use safer_ffi::prelude::*;
+
+use crate::ffi::NsError;
+
+use super::NsNoosphereContext;
+
+#[ffi_export]
+/// Get the DID that is assigned to the provided petname; note that the DID is
+/// the ID of the sphere, but in order to read the sphere you must resolve the
+/// DID to a CID, which tells you the version of the sphere to read.
+pub fn ns_sphere_petname_get(
+    noosphere: &NsNoosphereContext,
+    sphere_identity: char_p::Ref<'_>,
+    petname: char_p::Ref<'_>,
+    error_out: Option<Out<'_, repr_c::Box<NsError>>>,
+) -> Option<char_p::Box> {
+    todo!();
+}
+
+#[ffi_export]
+/// Assign a DID to a petname. This will overwrite a petname entry if one already exists
+/// with the given name (and reset the resolved CID, if any).
+pub fn ns_sphere_petname_set(
```

Comment:
The tuple-nature of the address book (`(id, petname) => cid`) doesn't lead to obvious FFI patterns. We should document outcome if petname/did are NULL (implementation of course TBD). I don't think I'd expect the petname to be the "key" here, but changing DID or petname are both valid use cases. Wonder what it'd look like if e.g.

```c
links = ns_sphere_links_get();
for (uint8_t i = 0; i < links.len; i++) {
  ns_link* link = links.array[i];
  printf("name: %s", ns_link_get_name(link));
  printf("did: %s", ns_link_get_did(link));
  ns_link_set_name(link, "alice");
}
```

Though may cause similar synchronization issues as separation of Fs from Context.

Reply:
Note that under the hood, names (like content) are contained by a HAMT. This data structure is convenient (and necessary) because it allows us operate on the data with map semantics, while also allowing us to avoid loading in the entire data structure just to read one key.

The implication here is that loading in the entire set of names could be quite expensive as a general purpose operation, especially if we are working on one key at a time. It's expensive in terms of memory allocation and it's expensive in terms of storage access because the key space of the "map" is effectively sharded across many blocks.

That having been said, even in cases where synchronization _is not_ an issue, we would still want to mediate access to the list of petnames via FFI. And, on the Rust side we implement iteration over all petnames/slugs/etc. as `Stream`s - which are effectively async generators - so that we can side-step the pathological case of loading the entire key space into memory at once.

So, I guess the takeaway here is that we need a FFI to access the full space of petnames.

Reply:
Well explained, thanks!